keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java

```diff
@@ -0,0 +1,93 @@
+package io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.unary;
+
+import io.improbable.keanu.annotation.ExportVertexToPythonBindings;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.LoadVertexParam;
+import io.improbable.keanu.vertices.SaveVertexParam;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.Differentiable;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivative;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+public class PermuteVertex extends DoubleUnaryOpVertex implements Differentiable {
+
+    private final static String REARRANGE_NAME = "rearrange";
+
+    private final int[] rearrange;
+
+    @ExportVertexToPythonBindings
+    public PermuteVertex(@LoadVertexParam(INPUT_VERTEX_NAME) DoubleVertex inputVertex,
+                         @LoadVertexParam(REARRANGE_NAME) int... rearrange) {
+        super(calculatePermutedShape(inputVertex, rearrange), inputVertex);
+        this.rearrange = rearrange;
+    }
+
+    @Override
+    protected DoubleTensor op(DoubleTensor value) {
+        return value.permute(rearrange);
+    }
+
+    @Override
+    public PartialDerivative forwardModeAutoDifferentiation(Map<Vertex, PartialDerivative> derivativeOfParentsWithRespectToInput) {
+        PartialDerivative derivativeOfParentWithRespectToInputs = derivativeOfParentsWithRespectToInput.get(inputVertex);
+        int[] permuteToApply = forwardPermute(derivativeOfParentWithRespectToInputs);
+        return derivativeOfParentWithRespectToInputs.permute(permuteToApply);
+    }
+
+    @Override
+    public Map<Vertex, PartialDerivative> reverseModeAutoDifferentiation(PartialDerivative derivativeOfOutputWithRespectToSelf) {
+        Map<Vertex, PartialDerivative> partials = new HashMap<>();
+        int[] permuteToApply = reversePermute(derivativeOfOutputWithRespectToSelf);
+        partials.put(inputVertex, derivativeOfOutputWithRespectToSelf.permute(permuteToApply));
+        return partials;
+    }
+
+    private int[] forwardPermute(PartialDerivative partial) {
+        long[] wrtShape = partial.getWrtShape(inputVertex.getShape());
+        int[] permuteToApply = new int[partial.get().getRank()];
+
+        for (int i = 0; i < rearrange.length; i++) {
+            permuteToApply[i] = rearrange[i];
+        }
+
+        for (int j = 0; j < wrtShape.length; j++) {
+            permuteToApply[j + rearrange.length] = j + rearrange.length;
+        }
+
+        return permuteToApply;
+    }
+
+    private int[] reversePermute(PartialDerivative partial) {
```

Comment:
The n^2 for loop is nasty and not needed. You can calculate what the inversion of the permute will be at construction time using something like:

```
...
   this.invertedRearrange = invertedPermute(rearrange);
...
    private static int[] invertedPermute(int[] rearrange) {
        int[] inverted = new int[rearrange.length];

        for (int i = 0; i < rearrange.length; i++) {
            inverted[rearrange[i]] = i;
        }

        return inverted;
    }
```
And then later just adjust it for the rank of the incoming partial:
```
    private int[] reversePermute(PartialDerivative partial) {

        int partialRank = partial.get().getRank();
        int[] permuteToApply = new int[partialRank];
        int ofRank = partialRank - this.getRank();

        for (int i = 0; i < partialRank; i++) {
            if (i >= ofRank) {
                permuteToApply[i] = invertedRearrange[i - ofRank] + ofRank;
            } else {
                permuteToApply[i] = i;
            }
        }

        return permuteToApply;
    }
```

Reply:
nice!